### DISCUSSION TITLE
Hydrating readonly atoms in jest tests

### DISCUSSION THREAD

**Discussion by avisprince** [pmndrs/jotai#2349] [Q&A]
I am following the testing documentation for writing my unit tests. However, I am running into errors with hydrating my readonly atoms. Can anyone point me to a solution? I am using the approach of wrapping my components in a provider when rendering my test:

```
const store = createStore()
store.set(atom1, value1)
store.set(readonlyAtom, value) // throws an error

<Provider store={store}>
  <ComponentBeingTested />
</Provider>
```

This is also happening when using `atomWithQuery`. I have also tried mocking the atoms I created directly, but jest doesn't do the best job mocking constants from files. So, while this approach worked as a global mock, I couldn't change the returned values between tests. I also tried using `jest.spyOn` on `useAtom` and then returning mocked values based on the atom's debugValue, but I get an error saying I cannot redefine the function so the mock fails. At this point, my only other idea is to make custom hooks wrapping the jotai hooks, but I don't want to have to do that.

Any help would be appreciated! Thank you!

**Comment by dai-shi** [comment]
I don't think of any ways without modifying runtime code, if mocking doesn't work for you case. 😓
